iree/compiler/Codegen/LLVMCPU/VectorContractCustomKernels.cpp

```diff
@@ -0,0 +1,317 @@
+    } else if (isReductionIterator(iteratorTypes[i])) {
+      reduction_iterators.push_back(i);
+    } else {
+      return false;
+    }
+  }
+  if (parallel_iterators.size() != 2 || reduction_iterators.size() != 1) {
+    return false;
+  }
+  // Give the found iterators some idiomatic names.
+  const int M = parallel_iterators[0];
+  const int N = parallel_iterators[1];
+  const int K = reduction_iterators[0];
+  // Check that there are 3 indexing maps.
+  auto indexingMaps = contractionOp.indexing_maps().getValue();
+  if (indexingMaps.size() != 3) {
+    return false;
+  }
+  // Check that the indexing maps have the expected form.
+  const int expectedMapResults[3][2] = {{M, K}, {N, K}, {M, N}};
+  for (int m = 0; m < 3; ++m) {
+    auto map = indexingMaps[m].cast<AffineMapAttr>().getValue();
+    if (map.getNumDims() != 3 || map.getNumResults() != 2) {
+      return false;
+    }
+    for (int r = 0; r < 2; ++r) {
+      int actualMapResult =
+          map.getResults()[r].cast<AffineDimExpr>().getPosition();
+      if (actualMapResult != expectedMapResults[m][r]) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+// Returns true if `contractionOp` is of the form
+//   matrix * transposed_matrix
+// where matrix is a vector<{mSize}x{kSize}xType>, and
+// transposed_matrix is a vector<{nSize}x{kSize}xType>
+static bool isMatrixTimesMatrixTransposed(vector::ContractionOp contractionOp,
+                                          int64_t mSize, int64_t kSize,
+                                          int64_t nSize) {
+  if (!isMatrixTimesMatrixTransposed(contractionOp)) {
+    return false;
+  }
+  VectorType lhsType = contractionOp.lhs().getType().cast<VectorType>();
+  VectorType rhsType = contractionOp.rhs().getType().cast<VectorType>();
+  auto lhsShape = lhsType.getShape();
+  auto rhsShape = rhsType.getShape();
+  if (lhsShape[0] != mSize || lhsShape[1] != kSize || rhsShape[0] != nSize ||
+      rhsShape[1] != kSize) {
+    return false;
+  }
+  return true;
+}
+
+// Checks that the Value `extResult` is defined by an arith::ExtSIOp promoting
+// from `extSrcType` to `extDstType`, and returns the input of the ExtSIOp.
+static Value getExtSIInput(Type extSrcType, Type extDstType, Value extResult) {
+  auto extSIOp = extResult.getDefiningOp<arith::ExtSIOp>();
+  if (!extSIOp) {
+    return nullptr;
+  }
+  Value extInput = extSIOp.in();
+  if (extInput.getType().cast<VectorType>().getElementType() != extSrcType) {
+    return nullptr;
+  }
+  return extInput;
+}
+
+// Helper to create a 1D, contiguous slice of a 1D vector.
+static Value extract1DSlice(PatternRewriter &rewriter, Location loc,
+                            VectorType dstVecType, Value input, int position) {
+  assert(input.getType().cast<VectorType>().getRank() == 1);
+  assert(dstVecType.getRank() == 1);
+  SmallVector<int64_t> offsets{position};
+  SmallVector<int64_t> strides{1};
+  return rewriter.create<vector::ExtractStridedSliceOp>(
```

Comment:
Done.
(Note, though, that all of this --- both ArrayRef vs SmallVector and even SmallVector vs std::vector --- is misguided optimization, because the kind of c++ code that we write in MLIR patterns is slow c++ for bigger, deeper reasons. It's Java-style C++, the primary bottleneck is probably the chains of pointer indirections due to the heavy runtime polymorphism. Clearly this C++ codebase has been designed with no concern for performance, which is fine, but then the small-scale optimizations such as SmallVector, ArrayRef, and all the use of explicit move semantics, come across as a misguided tradeoff, bringing in unsafety (e.g. taking references introduces risks of dangling references to dead temporaries, explicit moves introduce the risk of referencing an already moved-out object) in exchange for very hypothetical performance gains.

Reply:
Ah of course. I think if you pass them directly as a literal it works fine, but not if you extract them in to a variable.

To your more general comment, I think you're right that there are places where MLIR is optimizing unnecessarily, but the overall analysis that MLIR was not designed for performance is off the mark (and I know some people who would probably be offended at the aspersion :stuck_out_tongue:). It's not designed for performance in the same way that a hyper-optimized c runtime code is, but performance is very much a concern. In particular, great pains are taken to avoid the exact pointer indirection chains you're referring to. For instance, Operations are not subclasses, but smart pointers, classes are frequently defined with CRTP to avoid dynamic virtual method called, etc. You'll see quite a lot of profile-guided optimization being discussed in the discourse as well. Some of the optimization here is based on what I assume is past experience with LLVM, where I believe folks saw considerable advantages (hence its existence).